croncheck.php

```diff
@@ -121,8 +121,37 @@
     header('Expires: Tue, 04 Sep 2012 05:32:29 GMT');
 }
 
+if (isset($CFG->adminsetuppending)) {
+    send_critical("Admin setup pending, please set up admin account");
+}
+
 if (moodle_needs_upgrading()) {
-    send_critical("Moodle upgrade pending, cron execution suspended");
+    $upgraderunning = get_config(null, 'upgraderunning');
+    $initialinstall = during_initial_install();
+
+    if ($upgraderunning && $upgraderunning >= time()) {
+        // Before the expected finish time.
+        if (!empty($initialinstall)) {
+            send_critical("Moodle initial installation is running, cron execution suspended");
+        } else {
+            send_critical("Moodle upgrade is running, cron execution suspended");
+        }
+    } else if ($upgraderunning && $upgraderunning < time()) {
+        $difference = (time() - $upgraderunning) / 60;
+        /*
+         * After the expected finish time (timeout or other interruption)
+         * The "core_shutdown_manager::register_function('upgrade_finished_handler');" already handle these cases
+         * and unset config 'upgraderunning'
+         * The below critical ones can happen if core_shutdown_manager fails to run the handler function.
+         */
+        if (!empty($initialinstall)) {
+            send_critical("Moodle initial installation is running and overdue by $difference mins");
+        } else {
+            send_critical("Moodle upgrade is running and overdue by $difference mins");
+        }
+    } else {
+        send_critical("Moodle upgrade pending and is not running, cron execution suspended");
```

Comment:
If you move this test to the top it should un-nest the complex if and make it easier to read, ie

```
if (!$upgraderunning) {
   send_critical("Moodle upgrade pending and is not running, cron execution suspended");
}
```

then other checks afterwards don't need the $upgraderunning

if ($upgraderunning && whateverelse)

Reply:
Hi @brendanheywood, I have changed the code